### PR TITLE
Fix #4904 -  [Sync Integration] Bookmarks tests are failing

### LINKED ISSUE #4904: [Sync Integration] Bookmarks tests are failing
See Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1549422

Looks like something with TPS. Manually tests work but automatically it is not possible to find the created bookmark in any of the scenarios (Mobile <-> Desktop)

### PR DESCRIPTION
**Pull request by justindarc**
CC: @isabelrios 

Isabel, can you try testing this build using the build that comes from BB?

### CODE REVIEW

**Comment by isabelrios** [comment]
> Isabel, can you try testing this build using the build that comes from BB?

@justindarc Is there a difference if I try directly on your branch? I just tried and the bookmark does not appear :( in any of the folders...
I need to investigate how to run the tests in a given build, not sure if that's possible they way they are written...

**Comment by isabelrios** [comment]
Firefox logs attached. Unfortunately still unable to see the Bookmarks after sync  :( 
[firefox.log](https://github.com/mozilla-mobile/firefox-ios/files/3158487/firefox.log)

Thanks for your help with this!

**Comment by mhammond** [comment]
> Firefox logs attached. Unfortunately still unable to see the Bookmarks after sync :(
> [firefox.log](https://github.com/mozilla-mobile/firefox-ios/files/3158487/firefox.log)

That log still has
`1557334162593	Sync.Synchronizer	TRACE	The bookmarks engine was disabled remotely.`

Looking at desktop's impl, this appears to mean the engine isn't in the "engines" part of meta/global (ie, it's not about info/collections nor about declined)

**Comment by isabelrios** [comment]
@justindarc @mhammond, I have new info to share about this issue, hope this shed some light on the issue. After doing more testing, I got one of the tests working by changing the server. The tests use the stage server and were failing with that error, well, changing to production the error is gone and tests pass. 

I tried the scenario Desktop-> iOS and the bookmark was shown. The only thing is that it was inside `Mobile Bookmarks` folder, not `Unsorted Bookmarks`

I have to try also the other scenario iOS -> Desktop and Logins and I will let you know.

**Comment by isabelrios** [comment]
And the scenario Login created on Desktop checked on iOS also works that way...

File: Sync/SyncStateMachine.swift [hunk 1]
```diff
@@ -25,14 +25,18 @@ private let LocalEngines: [String] = TogglableEngines + ["clients"]
 // Names of collections which will appear in a default meta/global produced locally.
 // Map collection name to engine version.  See http://docs.services.mozilla.com/sync/objectformats.html.
 private let DefaultEngines: [String: Int] = [
+    "bookmarks": 2,
     "clients": ClientsStorageVersion,
     "history": HistoryStorageVersion,
     "tabs": TabsStorageVersion,
     // We opt-in to syncing collections we don't know about, since no client offers to sync non-enabled,
     // non-declined engines yet.  See Bug 969669.
```
Comment (linacambridge) [inline Sync/SyncStateMachine.swift]: @justindarc If you think it'll be helpful, would you mind cribbing [this comment](https://github.com/mozilla-mobile/firefox-ios/pull/4549/files#diff-996382204f85c037854c773b09b618d6R33), too?

**Comment by linacambridge** [comment]
> After doing more testing, I got one of the tests working by changing the server. The tests use the stage server and were failing with that error, well, changing to production the error is gone and tests pass.

That's really interesting! Would you mind sharing the TPS configs for both Desktop and iOS, @isabelrios? I was going to try to build Firefox for iOS locally, add some logging, and see if I could inspect the `meta/global` records by hand in both (iOS to Desktop, and Desktop to iOS) cases.

**Comment by isabelrios** [comment]
> Would you mind sharing the TPS configs for both Desktop and iOS

Hi @linacambridge is this what you mean by the TPS cofigs?: 
[iOS-> Desktop bookmark](https://github.com/mozilla-mobile/firefox-ios/blob/master/SyncIntegrationTests/test_bookmark.js)
[Desktop-> iOS bookmark](https://github.com/mozilla-mobile/firefox-ios/blob/master/SyncIntegrationTests/test_bookmark_desktop.js)
All the code and info about these tests is [here](https://github.com/mozilla-mobile/firefox-ios/tree/master/SyncIntegrationTests)
Let me know if you want me to share or test anything else. Thanks!

**Comment by linacambridge** [comment]
Hi @isabelrios! Not quite, although that's also helpful, thanks! I meant the `config.json` for TPS, that looks like [this](https://searchfox.org/mozilla-central/rev/b9da45f63cb567244933c77b2c7e827a057d3f9b/testing/tps/config/config.json.in).

**Comment by mhammond** [comment]
> After doing more testing, I got one of the tests working by changing the server.

So what might be happening is that TPS probably doesn't perform a full storage wipe before or after the tests. So the first time TPS shows up on a server for a given user, `meta/global` is written at that time, but subsequent TPS runs end up leaving it there.

That would explain why a test with this patch doesn't fix the issue - it's because the build with this patch didn't actually write meta/global. By shifting to a new server (or, I suspect, a new username), we wrote the correct meta/global and life was good.

I think we should just assume that's the case :) The fact this build is demonstrated to work with a new server is good enough for me to consider this resolved once this patch lands.

**Comment by isabelrios** [comment]
> By shifting to a new server (or, I suspect, a new username)

Using a new user does not fix this, for each test run a new user was created and removed after the tests were done. 
Would it be fine to use the prod server to run these tests daily? Not sure if creating a prod account each time tests run and deleting it would be something we want to avoid.

**Comment by isabelrios** [comment]
@linacambridge oh I see, then I think [this is](https://github.com/mozilla-mobile/firefox-ios/blob/master/SyncIntegrationTests/conftest.py#L59) what you want 

@mhammond I have not forgotten about your add-on, will try to share the info there by today.

**Comment by mhammond** [comment]
> Using a new user does not fix this, for each test run a new user was created and removed after the tests were done.

That's very odd and surprising.

> Would it be fine to use the prod server to run these tests daily?

We "got in trouble" for this running desktop TPS tests (although we may have been doing it more than once per day). I'm not sure how prod would fix this though - wouldn't it have the same problem?

**Comment by isabelrios** [comment]
> > Would it be fine to use the prod server to run these tests daily?
> 
> We "got in trouble" for this running desktop TPS tests (although we may have been doing it more than once per day). I'm not sure how prod would fix this though - wouldn't it have the same problem?

I don't really know if there is a difference between the two servers...but tests fail on stage and works on production...that's what I saw while running them under different scenarios. If you want we can schedule a zoom video to show you both cases or we can wait until Whistler and meet there

**Comment by isabelrios** [comment]
This issue is getting very long but so that all info is together I would like to add a few more findings....

I was able to use the [about-sync add-on ](https://addons.mozilla.org/en-US/firefox/addon/about-sync/) with an stage account and with nightly set up to use stage. 
These are the logs in both cases:
Bookmark saved on iOS -> checked on Desktop: 
[iOSDesktop.docx](https://github.com/mozilla-mobile/firefox-ios/files/3186679/iOSDesktop.docx)

Bookmark added to Desktop via TPS -> checked on iOS:
[DesktopiOS.docx](https://github.com/mozilla-mobile/firefox-ios/files/3186682/DesktopiOS.docx)

I don't see any error logs, they all are shown as successful...but I may be missing something or other logs..
@mhammond Are there other logs more useful to get from the add on? Let me know so that I can share...

Also, I would like to add that it is not only something about the tests. I set iOS sim to use stage and logged in with the sync account used and the bookmark was not shown. So it is not a timing issue in the test itself to wait for the bookmark enough time.

**Comment by mhammond** [comment]
> This issue is getting very long but so that all info is together I would like to add a few more findings....

The good news is that the issue this PR was trying to fix appears to have been fixed - we are no longer seeing the bookmarks engine as having been remotely disabled - so we should probably close this PR.

The bad news is that TPS still isn't working for you :(

Should we make a TPS slack channel?

> 
> I was able to use the [about-sync add-on ](https://addons.mozilla.org/en-US/firefox/addon/about-sync/) with an stage account and with nightly set up to use stage.
> These are the logs in both cases:
> Bookmark saved on iOS -> checked on Desktop:
> [iOSDesktop.docx](https://github.com/mozilla-mobile/firefox-ios/files/3186679/iOSDesktop.docx)

That log is a little strange - it shows us syncing due to a push message, but because it's from a push message, bookmarks aren't synced. I suspect the "interesting" logs are in a different success log file on the same profile (although I suspect there actually aren't any "interesting" logs at all, and that desktop simply isn't seeing the iOS created bookmark)

> Bookmark added to Desktop via TPS -> checked on iOS:
> [DesktopiOS.docx](https://github.com/mozilla-mobile/firefox-ios/files/3186682/DesktopiOS.docx)

This log shows bookmarks being synced, but they found nothing to do (ie, no new stuff to be downloaded from the server, and no local changes to be uploaded to the server) - but again, this might not be the correct log.

> I don't see any error logs, they all are shown as successful...but I may be missing something or other logs..
> @mhammond Are there other logs more useful to get from the add on? Let me know so that I can share...

Sync will create a new log file for every single sync, so it can be difficult to work out which of the many log files relate to the specific TPS problem. In general, it's best to capture the logs from TPS, as those logs have *both* the sync logs *and* the TPS logs in one convenient file, so it's almost impossible to choose the wrong log file.
 
> Also, I would like to add that it is not only something about the tests. I set iOS sim to use stage and logged in with the sync account used and the bookmark was not shown. So it is not a timing issue in the test itself to wait for the bookmark enough time.

If you've run the iOS portion of the test and then manually start a desktop profile connected to the same account, then you certainly should see the iOS created bookmark appear in desktop. about:sync on desktop will also let you see the server records, so if you can't find the bookmark there either, then it points to the iOS portion of the test either not creating the bookmark, or not syncing it correctly after it is created.

Do we have a way of seeing any logs at all from an iOS run?

**Comment by isabelrios** [comment]
> In general, it's best to capture the logs from TPS, as those logs have both the sync logs and the TPS logs in one convenient file, so it's almost impossible to choose the wrong log file.

Can I get those TPS logs from the add-on? 

> If you've run the iOS portion of the test and then manually start a desktop profile connected to the same account, then you certainly should see the iOS created bookmark appear in desktop. about:sync on desktop will also let you see the server records, so if you can't find the bookmark there either, then it points to the iOS portion of the test either not creating the bookmark, or not syncing it correctly after it is created.

The same thing happens the other way around, if I create the bookmark on desktop manually and then try to check on iOS. The bookmark is not there. I'm not saying it is not iOS fault, just to have more info :)

> Do we have a way of seeing any logs at all from an iOS run?

The logs we have from that part are xcode logs (just looking at the UI) so there is not enough info only that the bookmark is now there. Honestly I don't know how to add more logs to look deeper... @justindarc any idea? Thanks!

**Pull request merged**